rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+RandomizedDelaySec=90
+
+[Install]
+WantedBy=teleport-agent.service
+```
+
+The exact time values are TBD. We want upgrades to be reliably and frequently discovered, but don't want to end up overloading
+repositories. This is why we aren't just using `OnActiveSec=1`.  If teleport is in *too* tight of a crash loop, we don't want
+to just end up hitting the repos in a tight loop. This will definitely take some fine-tuning, and we may want to customize the
+`StartLimitIntervalSec` and `StartLimitBurst` of the main teleport agent unit.
+
+---
+
+TODO: apt/yum channel and artifact promotion details
+
+
+### Automatic/Guided Install Changes
+
+Teleport provides two kinds of automatic/guided install mechanism for adding the teleport agent software to existing
+machines. There are "discovery installers" which are scripts used by ec2 auto-discovery (and possibly by other forms of
+server auto-discovery in the future) to install teleport agents automatically, and there is the `node-join` script template
+used by the web UI to serve custom installers during the interactive agent-adding process.
+
+To ensure a seamless experience, we will need both of these install scripts to be capable of being parameterized or overridden s.t.
+they target `cloud-stable` when the cluster is configured for continuous agent upgrades.
+
+The current implementation of the `node-join` script template relies upon direct downloads from `get.gravitational.com`. It would
+be preferable to force use of a package manager when installing `cloud-stable`, as that would let us improve standardization, minimize
+the amount of custom logic in the download path, and (most importantly) reduce the complexity of the periodic upgrade operation,
+which is the most sensitive part of this proposal.
+
+I believe the best possible outcome would be to unify the core logic of the "discovery", "join", and "upgrade" scripts into a single
+cross-platform function so that we have a single source of truth for package manager discovery and invocation going forward. The
+feasability of this will require further investigation.
+
+
+### Summary of Changes
+
+- Agents will expose a flags to enable the VM and k8s variants of the agent upgrade behavior.
+
+- Package repositories will expose a `cloud-stable` variant of packages which will lag slightly behind "latest"
+and be managed indipendently of the standard release process.
+
+- A new `cloud-stable` helm chart (or option) will be dstributed.
+
+- A new `teleport-upgrade` script and accompanying systemd timer unit will be distributed as part of the
+`cloud-stable` apt/yum packages.
+
+- Teleport auth servers will be modified to be able to serve the maintenance window to agents.
+
+- Cloud clusters will have the option to track `cloud-stable` rather than a specific major version.
+
+- Teleport discovery and join features will be modified to serve appropriate install scripts when `cloud-stable`
+tracking is enabled for a cluster.
+
```

Comment:
I want to bring up the business/non-technical impact of automatic updates. I realize that we've already committed to automatic updates, which have many advantages, but I want to voice my concerns about the potential negative impact of doing so.

Automatic upgrades are risky. Customers rely on Teleport not just for day to day operations, but in many cases emergency access to systems. If automatic upgrades break and cause an outage then customers will be upset at Gravitational, even if they freely choose to enroll in automatic updates. If automatic upgrades break and don't cause an outage, but prevent customers from handling another costly issue in a timely manner, the finger will still get pointed at us.

I can't speak for cloud team as I am not really involved with our cloud offering, but I believe I can speak for our internal tooling team when I say that new releases can be hard. The v12 release took several PRs and fixes to get the release itself out the door. While the v12 release issues were largely based around our Drone pipelines, we could very well have (and previously have had) issues that affect our release binaries' ability to execute and run properly on client's machines.

It is my belief that if we screw up the automatic update process, or push out bad updates that either break customer systems, or prevent access to customer systems, customers place the blame on us. I believe if they have significant enough issues they will:
* Stop using automatic updates, defeating the purpose of this RFD,
* Stop using our cloud service, or
* Entirely leave us as customers

In my opinion the best way to prevent this outside of shear luck is to implement not just what is best or fastest for Gravitational, but what is best for the people (ops/system administrators) impacted by this proposal. I realize that we only have a few weeks to complete the project, but I am concerned that if we don't take this route there may be severe repercussions in the long term.

Reply:
I generally agree that automatic upgrades are huge risk.  FWIW we don't expect this feature to be enabled for all cloud customers.  Enabling this behavior will be done on a per-cluster basis (updated the intro to better communicate this), and we expect that we'll want to add additional more customizable upgrade features in the future.

> [...] implement not just what is best or fastest for Gravitational, but what is best for the people (ops/system administrators) impacted by this proposal.

Could you be more specific?  Assuming the target audience of this feature is only that subset of folks that are on board with the general concept of automatic upgrades, do you think we're failing to do what's best for _that_ target audience?  If so, how?  What do you think they'd prefer?

Reply:
Totally agree with @fheinecke
There's already some ways to implement automatic upgrades as part of the distro's pacakge management
https://wiki.debian.org/UnattendedUpgrades
https://dnf.readthedocs.io/en/latest/automatic.html

This could save the burden of doing manual upgrades by our customers.
There's still missing some pieces like:
- downgrade a version (can we do it using some repo's configuration on out side?)
- window upgrades per customer